imagecraft/ubuntu_image.py

```diff
@@ -18,13 +18,101 @@
 
 import subprocess
 
+from craft_application.util import dump_yaml
 from craft_cli import emit
+from pydantic import BaseModel, Field
 
 from imagecraft.errors import UbuntuImageError
 
 
-def generate_legacy_def_rootfs(  # noqa: PLR0913
+def _alias_generator(s: str) -> str:
+    return s.replace("_", "-")
+
+
+class Snap(BaseModel):
+    """Pydantic model for the Snap object in an ImageDefinition."""
+
+    name: str | None
+
+
+class Package(BaseModel):
+    """Pydantic model for the Package object in an ImageDefinition."""
+
+    name: str | None
+
+
+class Customization(BaseModel):
+    """Pydantic model for the Customization object in an ImageDefinition."""
+
+    extra_snaps: list[Snap] | None
+    extra_packages: list[Package] | None
+
+    class Config:
+        """Pydantic model configuration."""
+
+        validate_assignment = True
+        allow_mutation = True
+        allow_population_by_field_name = True
+        alias_generator = _alias_generator
+
+
+class Seed(BaseModel):
+    """Pydantic model for the Seed object in an ImageDefinition."""
+
+    urls: list[str]
+    branch: str
+    names: list[str]
+    pocket: str
+
+    class Config:
+        """Pydantic model configuration."""
+
+        validate_assignment = True
+        allow_mutation = True
+        allow_population_by_field_name = True
+        alias_generator = _alias_generator
+
+
+class Rootfs(BaseModel):
+    """Pydantic model for the Rootfs object in an ImageDefinition."""
+
+    components: list[str]
+    seed: Seed
+
+    class Config:
+        """Pydantic model configuration."""
+
+        validate_assignment = True
+        allow_mutation = True
+        allow_population_by_field_name = True
+        alias_generator = _alias_generator
+
+
+class ImageDefinition(BaseModel):
```

Comment:
I like this idea! I'm pretty sure we discussed that at some point, but have you looked into if it would be possible to somehow have some common, 'shared' schema for both ubuntu-image and imagecraft? You know, to have a single source of truth for the image definition schema? Sounds like non-trivial, but maybe..? This is good so far though!

Reply:
Oh yeah that is the first idea I had! :D I did this kind of things at my previous job, using protobuf. So I had a single set of protobuf files, and used the generator to derive the model in different languages. This is very handy to make sure models are kept up to date everywhere but it comes with a non-negligible overhead and complexity. I also thought about looking into tools that could convert a go struct to a pydantic class. This is more specify but could be easier to use considering our use case. 

Our image definition model is not evolving a lot, so I am not sure this is worth the investment though. I will keep it on my todo list for now and see if it becomes increasingly needed.